test/relayserver/ServerTestEnvironments.ts

```diff
@@ -0,0 +1,233 @@
+import { ethers as hardhat } from 'hardhat';
+import { expect } from 'chai';
+import { deployRelayHub, createSmartWalletFactory } from '../TestUtils';
+import config from 'config';
+
+const provider = hardhat.provider;
+const dayInSec = 24 * 60 * 60;
+const weekInSec = dayInSec * 7;
+
+const newServerInstance = async (
+  relayOwner: Wallet,
+  template: IForwarder,
+  serverWorkdirs?: ServerWorkdirs
+) => {
+  const relayServer = await newServerInstanceNoInit(
+    relayOwner,
+    template,
+    serverWorkdirs
+  );
+  await relayServer.init();
+  const latestBlock = await provider.getBlock('latest');
+  const receipts = await relayServer._worker(latestBlock.number);
+  await assertEventHub('RelayServerRegistered', receipts, ''); // sanity check
+  await relayServer._worker(latestBlock.number + 1);
+
+  return relayServer;
+};
+
+const newServerInstanceNoInit = async (
+  relayOwner: Wallet,
+  template: IForwarder,
+  serverWorkdirs?: ServerWorkdirs,
+  unstakeDelay = weekInSec
+) => {
+  const relayServer = await newServerInstanceNoFunding(
+    template,
+    serverWorkdirs
+  );
+
+  const { relayHubAddress, relayManagerAddress } = relayServer.pingHandler();
+
+  await relayOwner.sendTransaction({
+    value: utils.parseEther('2'),
+  });
+
+  const relayHub = RelayHub__factory.connect(relayHubAddress, provider);
+
+  await relayHub
+    .connect(relayOwner)
+    .stakeForAddress(relayManagerAddress, unstakeDelay, {
+      value: utils.parseEther('1'),
+    });
+
+  return relayServer;
+};
+
+const newServerInstanceNoFunding = async (
+  template: IForwarder,
+  serverWorkdirs?: ServerWorkdirs
+) => {
+  const { workdir } = getTemporaryWorkdirs();
+
+  const managerKeyManager = createKeyManager(serverWorkdirs?.managerWorkdir);
+  const workersKeyManager = createKeyManager(serverWorkdirs?.workersWorkdir);
+  const txStoreManager = new TxStoreManager({
+    workdir: serverWorkdirs?.workdir ?? workdir,
+  });
+
+  const dependencies: ServerDependencies = {
+    txStoreManager,
+    managerKeyManager,
+    workersKeyManager,
+  };
+
+  const factory = await createSmartWalletFactory(template);
+  const relayHub = await deployRelayHub(undefined, {});
+  const deployVerifierFactory = await hardhat.getContractFactory(
+    'DeployVerifier'
+  );
+  const deployVerifier = await deployVerifierFactory.deploy(factory.address);
+  const relayVerifierFactory = await hardhat.getContractFactory(
+    'RelayVerifier'
+  );
+  const relayVerifier = await relayVerifierFactory.deploy(factory.address);
+
+  const originalConfig = { ...config };
+
+  config.util.extendDeep(originalConfig, {
+    contracts: {
+      relayHubAddress: relayHub.address,
+      deployVerifierAddress: deployVerifier.address,
+      relayVerifierAddress: relayVerifier.address,
+      smartWalletFactoryAddress: factory.address,
+    },
+    app: {
+      devMode: true,
+      checkInterval: 10,
+      logLevel: 5,
+    },
+  });
+
+  const relayServer = new RelayServer(dependencies);
```

Comment:
We cannot set the provider from the server, it get the provider from ´getDefaultProvider´, but we need to use the hardhat one.

Reply:
We may need to use a similar approach to what we used in the client.

Reply:
I think that we can stub the getProvider from the server, to change that, but I need to test it.

Reply:
I was able to stub it.

```typescript
Sinon.stub(ethers, 'getDefaultProvider').returns(hardhat.provider);
```

Should I stub it, or create a function to setup it?

Reply:
I think we could stub it for now

Reply:
Ill stub it, during the process of the getInstance from the server.

Reply:
updated